drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,466 @@
+		uint8_t value_tach_target_count_msb = pulse_count >> 3;
+		uint8_t value_tach_target_count_lsb = (pulse_count & 0x05) << 5;
+
+		value_fan_configuration |= MAX37190_FANXCONFIGURATION_MODE_BIT;
+
+		result = max31790_write_registers(
+			dev, MAX31790_REGISTER_TACHTARGETCOUNTMSB(channel),
+			value_tach_target_count_msb, value_tach_target_count_lsb);
+		if (result != 0) {
+			return result;
+		}
+	}
+
+	result = max31790_write_register(dev, MAX37190_REGISTER_FANCONFIGURATION(channel),
+					 value_fan_configuration);
+	if (result != 0) {
+		return result;
+	}
+
+	result = max31790_write_register(dev, MAX31790_REGISTER_FANDYNAMICS(channel),
+					 value_fan_dynamics);
+	if (result != 0) {
+		return result;
+	}
+
+	return 0;
+}
+
+static int max31790_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_count,
+			       uint32_t pulse_count, pwm_flags_t flags)
+{
+	struct max31790_data *data = dev->data;
+	int result;
+
+	LOG_DBG("set period %i with pulse %i for channel %i and flags 0x%04X", period_count,
+		pulse_count, channel, flags);
+
+	if (channel > MAX31790_CHANNEL_COUNT) {
+		LOG_ERR("invalid channel number %i", channel);
+		return -EINVAL;
+	}
+
+	if (period_count == 0) {
+		LOG_ERR("period count must be > 0");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	result = max31790_set_cycles_internal(dev, channel, period_count, pulse_count, flags);
+	k_mutex_unlock(&data->lock);
+	return result;
+}
+
+static int max31790_get_cycles_per_sec(const struct device *dev, uint32_t channel, uint64_t *cycles)
+{
+	struct max31790_data *data = dev->data;
+	int result;
+	bool success;
+	uint8_t pwm_frequency_register;
+	uint8_t pwm_frequency = 1;
+	uint16_t pwm_frequency_in_hz;
+
+	if (channel > MAX31790_CHANNEL_COUNT) {
+		LOG_ERR("invalid channel number %i", channel);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	result = max31790_read_register(dev, MAX37190_REGISTER_GLOBALCONFIGURATION,
+					&pwm_frequency_register);
+
+	if (result != 0) {
+		k_mutex_unlock(&data->lock);
+		return result;
+	}
+
+	pwm_frequency = max31790_get_pwmfrequency(pwm_frequency_register, channel);
+	success = max31790_convert_pwm_frequency_into_hz(&pwm_frequency_in_hz, pwm_frequency);
+
+	if (!success) {
+		k_mutex_unlock(&data->lock);
+		return -EINVAL;
+	}
+
+	*cycles = pwm_frequency_in_hz;
+
+	k_mutex_unlock(&data->lock);
+	return 0;
+}
+
+static const struct pwm_driver_api max31790_api = {
+	.set_cycles = max31790_set_cycles,
+	.get_cycles_per_sec = max31790_get_cycles_per_sec,
+};
+
+static int max31790_init(const struct device *dev)
+{
+	const struct max31790_config *config = dev->config;
+	struct max31790_data *data = dev->data;
+	int result;
+	uint8_t register_value;
+
+	k_mutex_init(&data->lock);
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	result =
+		max31790_read_register(dev, MAX37190_REGISTER_GLOBALCONFIGURATION, &register_value);
```

Comment:
can you indent on the arguments?

Reply:
I could, but clang-format will change it back to this style. Therefore I would like to keep it like this, having code automatically formatted is preferable over manual formatting.

Reply:
Fair point, how about using something shorter for `register_value` then? Typically Zephyr drivers use `reg_val`, `val` or `value`.

Reply:
done